### PR TITLE
LBSearch previously wasn't always returning items that should have...

### PR DESCRIPTION
**Pull request by pjshumphreys**
been found; switched from upper bound address to base+offset algorithm

### CODE REVIEW

**Comment by pjshumphreys** [comment]
I have some sample code that shows how the previous code could fail at https://www.z88dk.org/forum/viewtopic.php?f=12&t=11681

File: libsrc/stdlib/Lbsearch.asm [hunk 1]
```diff
@@ -24,57 +24,42 @@ EXTERN l_jpix
...
-   rr h
-   rr l             ; HL = unrounded address of middle item
-
-   ld a,l           ; shenanigans to ensure HL aligns on item
-   xor e
-   rra
-   jr nc, compare
-   dec hl
+   push hl      ; Get the address at the halfway index + the base address
+   ld a, $fe
```
Comment (ped7g) [inline libsrc/stdlib/Lbsearch.asm:38]: is this for Z80? If yes, you can do also `res 0,l` to make HL even in 8T and two bytes.

Comment (pjshumphreys) [reply libsrc/stdlib/Lbsearch.asm:38]: It will need to work on an 8080 as well as on a z80 I think

Comment (pjshumphreys) [reply libsrc/stdlib/Lbsearch.asm:38]: Then again, the "rr h" instruction that was here previously isn't available on an 8080 either, so this function has never been 8080 friendly. I'll probably change the code to your suggestion to save a few T states. Every T state will count if this code is to search through large arrays

Comment (suborb) [reply libsrc/stdlib/Lbsearch.asm:38]: https://github.com/z88dk/z88dk/wiki/Classic-8080#stdlibh - bsearch is not available on 808x

Comment (pjshumphreys) [reply libsrc/stdlib/Lbsearch.asm:38]: This is a change to the l_bsearch function in stdlib. z88dk doesn't have bsearch at all it seems. The sample code I posted on the forums used a preprocessor macro to filter out the extra size parameter that bsearch has but that l_bsearch doesn't. bsearch on gcc can operate on arrays of pointers the same as l_bsearch (only) can.

Comment (pjshumphreys) [reply libsrc/stdlib/Lbsearch.asm]: Ok @ped7g, I've run my tests and the opcode suggested works fine.

**Comment by suborb** [comment]
Test to follow in next commit

**Pull request merged**